### ISSUE TITLE
Password entry not shown

### ISSUE THREAD

**Issue by Hoempi** [greenpau/caddy-security#89]
**Describe the issue**

I started with trying to mix the different examples from readme.md and /assets/config/Caddyfile to match my use case. But after entering the username in the portal the next screen shows "Unauthorized". This is happening regardless whether I use a username found in users.json or some random characters.

![grafik](https://user-images.githubusercontent.com/40570297/163352713-84031cb7-9358-418e-b7b1-efab19f70a2b.png)

**Configuration**

Paste full `Caddyfile` below:

```
{
	security {

		local identity store localdb {
			realm local
			path assets/config/users.json
		}

		authentication portal myportal {
			crypto default token lifetime 3600
      crypto key sign-verify {env.JWT_SECRET}			
			enable identity store localdb
			ui {
				links {
					"My Website" "/app" icon "las la-star"
					"My Identity" "/auth/whoami" icon "las la-user"
				}
			}
			transform user {
				match origin local
				action add role authp/user
				ui link "Portal Settings" /auth/settings icon "las la-cog"
			}
		}

		authorization policy mypolicy {
			set auth url /auth/
      crypto key verify {env.JWT_SECRET}
			allow roles authp/admin authp/user
		}
	}
}

172.16.0.50, example.myds.me  {

	route /version* {
		respond * "1.0.0" 200
	}

	route /auth* {
		authenticate * with myportal
	}

	route /xauth* {
		authenticate * with myportal
	}

	route /app {
		authorize with mypolicy
		file_server {
			root ./assets/config
		}
	}

	route {
		redir https://{hostport}/auth 302
	}
}

backup.example.myds.me {

  reverse_proxy 172.16.0.55:8001

}

dms.example.com {

  tls internal
  reverse_proxy 172.16.0.50:10807

}


dms.example.myds.me {

  reverse_proxy 172.16.0.50:10807

}

dsm.example.com, example.com {

  tls internal
  reverse_proxy *  {

    to 172.16.0.50:5001

    transport http {

      tls
      tls_insecure_skip_verify
    }

  }

}

dsm.example.myds.me, home.example.com

  reverse_proxy *  {

    to 172.16.0.50:5001

    transport http {

      tls
      tls_insecure_skip_verify
    }

  }

}

firefly.example.com {

  tls internal
  reverse_proxy 172.16.0.50:10824

}

firefly.example.myds.me {

  reverse_proxy 172.16.0.50:10824

}

jellyfin.example.com {

  tls internal
  reverse_proxy 172.16.0.50:10812

}

jellyfin.example.myds.me {

  reverse_proxy 172.16.0.50:10812

}

keycloak.example.com {

  tls internal
  reverse_proxy *  {

    to 172.16.0.50:10914

    transport http {

      tls
      tls_insecure_skip_verify
    }

  }

}

keycloak.example.myds.me {

  reverse_proxy *  {

    to 172.16.0.50:10914

    transport http {

      tls
      tls_insecure_skip_verify
    }

  }

}

medien.example.com {

  tls internal
  reverse_proxy 172.16.0.50:10822

}

medien.example.myds.me {

  reverse_proxy 172.16.0.50:10822

}


monica.example.com {

  tls internal
  reverse_proxy 172.16.0.50:10813

}

monica.example.myds.me {

  reverse_proxy 172.16.0.50:10813

}

mtv.example.com {

  tls internal
  reverse_proxy 172.16.0.50:10810

}

mtv.example.myds.me {

  reverse_proxy 172.16.0.50:10810

}



prism.example.com {

  tls internal
  reverse_proxy 172.16.0.50:10818

}

pihole.example.myds.me {

  reverse_proxy 172.16.0.60:80

}

plex.example.com {

  tls internal
  reverse_proxy 172.16.0.50:32400

}

portainer.example.com {

  tls internal
  reverse_proxy 172.16.0.50:10801

}

portainer.example.myds.me {

  reverse_proxy 172.16.0.50:10801

}

rss.example.com {

  tls internal
  reverse_proxy 172.16.0.50:10805

}

rss.example.myds.me {

  reverse_proxy 172.16.0.50:10805

}

speed.example.com {

  tls internal
  reverse_proxy 172.16.0.50:10811

}

speed.example.myds.me {

  reverse_proxy 172.16.0.50:10811

}

tmm.example.com {

  tls internal
  reverse_proxy 172.16.0.50:10816

}

tmm.example.myds.me {

  reverse_proxy 172.16.0.50:10816

}

torrent.example.com {

  tls internal
  reverse_proxy 172.16.0.50:10815

}


torrent.example.myds.me {

  reverse_proxy 172.16.0.50:10815

}

vault.example.com {

  tls internal
  reverse_proxy 172.16.0.50:10809

}

vault.example.myds.me {

  reverse_proxy 172.16.0.50:10809

}


webtrees.example.com {

  tls internal
  reverse_proxy 172.16.0.50:10817

}

webtrees.example.myds.me {

  reverse_proxy 172.16.0.50:10817

}

```

**Version Information**

Provide output of `caddy list-modules -versions | grep git` below:

```
admin.api.load v2.4.6
admin.api.metrics v2.4.6
admin.api.reverse_proxy v2.4.6
caddy.adapters.caddyfile v2.4.6
caddy.config_loaders.http v2.4.6
caddy.listeners.tls v2.4.6
caddy.logging.encoders.console v2.4.6
caddy.logging.encoders.filter v2.4.6
caddy.logging.encoders.filter.delete v2.4.6
caddy.logging.encoders.filter.ip_mask v2.4.6
caddy.logging.encoders.filter.replace v2.4.6
caddy.logging.encoders.json v2.4.6
caddy.logging.encoders.single_field v2.4.6
caddy.logging.writers.discard v2.4.6
caddy.logging.writers.file v2.4.6
caddy.logging.writers.net v2.4.6
caddy.logging.writers.stderr v2.4.6
caddy.logging.writers.stdout v2.4.6
caddy.storage.file_system v2.4.6
http v2.4.6
http.authentication.hashes.bcrypt v2.4.6
http.authentication.hashes.scrypt v2.4.6
http.authentication.providers.http_basic v2.4.6
http.encoders.gzip v2.4.6
http.encoders.zstd v2.4.6
http.handlers.acme_server v2.4.6
http.handlers.authentication v2.4.6
http.handlers.encode v2.4.6
http.handlers.error v2.4.6
http.handlers.file_server v2.4.6
http.handlers.headers v2.4.6
http.handlers.map v2.4.6
http.handlers.metrics v2.4.6
http.handlers.push v2.4.6
http.handlers.request_body v2.4.6
http.handlers.reverse_proxy v2.4.6
http.handlers.rewrite v2.4.6
http.handlers.static_response v2.4.6
http.handlers.subroute v2.4.6
http.handlers.templates v2.4.6
http.handlers.vars v2.4.6
http.matchers.expression v2.4.6
http.matchers.file v2.4.6
http.matchers.header v2.4.6
http.matchers.header_regexp v2.4.6
http.matchers.host v2.4.6
http.matchers.method v2.4.6
http.matchers.not v2.4.6
http.matchers.path v2.4.6
http.matchers.path_regexp v2.4.6
http.matchers.protocol v2.4.6
http.matchers.query v2.4.6
http.matchers.remote_ip v2.4.6
http.matchers.vars v2.4.6
http.matchers.vars_regexp v2.4.6
http.precompressed.br v2.4.6
http.precompressed.gzip v2.4.6
http.precompressed.zstd v2.4.6
http.reverse_proxy.selection_policies.cookie v2.4.6
http.reverse_proxy.selection_policies.first v2.4.6
http.reverse_proxy.selection_policies.header v2.4.6
http.reverse_proxy.selection_policies.ip_hash v2.4.6
http.reverse_proxy.selection_policies.least_conn v2.4.6
http.reverse_proxy.selection_policies.random v2.4.6
http.reverse_proxy.selection_policies.random_choose v2.4.6
http.reverse_proxy.selection_policies.round_robin v2.4.6
http.reverse_proxy.selection_policies.uri_hash v2.4.6
http.reverse_proxy.transport.fastcgi v2.4.6
http.reverse_proxy.transport.http v2.4.6
pki v2.4.6
tls v2.4.6
tls.certificates.automate v2.4.6
tls.certificates.load_files v2.4.6
tls.certificates.load_folders v2.4.6
tls.certificates.load_pem v2.4.6
tls.certificates.load_storage v2.4.6
tls.handshake_match.remote_ip v2.4.6
tls.handshake_match.sni v2.4.6
tls.issuance.acme v2.4.6
tls.issuance.internal v2.4.6
tls.issuance.zerossl v2.4.6
tls.stek.distributed v2.4.6
tls.stek.standard v2.4.6

  Standard modules: 83

http.authentication.providers.authorizer v1.1.3
http.handlers.authenticator v1.1.3
http.handlers.trace v1.1.8
security v1.1.3

  Non-standard modules: 4

  Unknown modules: 0
```

**Expected behavior**

I am expecting to be asked for the password.

**Additional context**

I do not use webadmin but provided AUTHP_ADMIN_USER, AUTHP_ADMIN_EMAIL, AUTHP_ADMIN_SECRET. Also I am running Caddy Security in Docker, using delver/caddy-security

Of course example.com is not the domain in question and so on. I used example as an example for anonymizing.

**Comment by greenpau** [comment]
>  I am running Caddy Security in Docker, using delver/caddy-security

@Hoempi , try running this solution with this image https://github.com/greenpau/caddy-security/pkgs/container/authp and let me know the outcome.

1. Are you using HTTP or HTTPS for accessing the portal?

2. Have you considered setting path to `/assets/config/users.json`?

```
		local identity store localdb {
			realm local
			path assets/config/users.json
		}
```

---

Side note: Have you considered using snippets?

```
mtv.example.com {

  tls internal
  reverse_proxy 172.16.0.50:10810

}

mtv.example.myds.me {

  reverse_proxy 172.16.0.50:10810

}
```

would be something like:

```
(domainHostTemplate) {
  {args.0}.com {
    tls internal
    reverse_proxy {args.1}
  }

  {args.0}.myds.me {
    reverse_proxy {args.1}
  }
}

import domainHostTemplate mtv.example 172.16.0.50:10810
import domainHostTemplate speed.example 172.16.0.50:10810
```

Also, please perform `caddy format` on your config and amend it here. It is pretty hard to read.

**Comment by greenpau** [comment]
@Hoempi , in case you run HTTP:

```
cookie insecure <on|off>: Allows sending cookies over HTTP. By default, it is disabled.
```

**Comment by Hoempi** [comment]
First of all: thanks for the quick reply and all the pointers. I'm pretty new to Caddy.

> > I am running Caddy Security in Docker, using delver/caddy-security
> 
> @Hoempi , try running this solution with this image https://github.com/greenpau/caddy-security/pkgs/container/authp and let me know the outcome.
Unfortunately still the same.
 
>     1. Are you using HTTP or HTTPS for accessing the portal?
HTTPS

>     2. Have you considered setting path to `/assets/config/users.json`?

If you mean in my docker compose file, yes:
`- /volume1/docker/caddy/assets:/srv/assets`
And like mentioned my predefined username shows up there.

> 
> ```
> 		local identity store localdb {
> 			realm local
> 			path assets/config/users.json
> 		}
> ```
> 
> Side note: Have you considered using snippets?
I was not aware of them, I'll definitely will look into it. Thanks a lot!

> 
> ```
> mtv.example.com {
> 
>   tls internal
>   reverse_proxy 172.16.0.50:10810
> 
> }
> 
> mtv.example.myds.me {
> 
>   reverse_proxy 172.16.0.50:10810
> 
> }
> ```
> 
> would be something like:
> 
> ```
> (domainHostTemplate) {
>   {args.0}.com {
>     tls internal
>     reverse_proxy {args.1}
>   }
> 
>   {args.0}.myds.me {
>     reverse_proxy {args.1}
>   }
> }
> 
> import domainHostTemplate mtv.example 172.16.0.50:10810
> import domainHostTemplate speed.example 172.16.0.50:10810
> ```
> 
> Also, please perform `caddy format` on your config and amend it here. It is pretty hard to read.
The result is
`/srv # caddy format
[ERROR] 'format' is not a recognized subcommand; see 'caddy help'`
Also I could not find a format directive in help and the only mention I found at caddyserver.com was https://caddyserver.com/docs/caddyfile/directives/log#format-modules

**Comment by greenpau** [comment]
> [ERROR] 'format' is not a recognized subcommand; see 'caddy help'`

Try:

```
caddy fmt -overwrite path/to/Caddyfile
caddy validate -config path/to/Caddyfile
caddy run -config path/to/Caddyfile
```

**Comment by greenpau** [comment]
> And like mentioned my predefined username shows up there. Might this be the problem? Did I mask stuff by not just mapping the file but the whole directory?

idk.

**Comment by Hoempi** [comment]
> caddy fmt -overwrite path/to/Caddyfile

I did that, but only some empty lines were deleted, nothing else.

I just got a RPI3 setup with DietPi and Docker Compose. So I will try to start from scratch on that one with only the stuff for the auth portal itself and then adding the other services step by step.

I will get back to you, as soon as I found out something or have a new Caddyfile which works. Thans for the help so far, really appreciate it!

**Comment by Hoempi** [comment]
Okay, moving to that RaspberryPi did the trick. I'm now getting the prompt and am able to log in. I got the notion because before I was running it on my Synology NAS and I already stumbled about other projects mentioning that some stuff inside Docker on Synology is not working as it should.

Once I get my configuration on the RasPi to my liking I will test that config on the NAS again, just to make sure I did not have anything else.

I've got another question, but I will open up a seperate issue for that, if I do not find anything in the doumentation or the other issues.
